jormungandr: Error handling in stat_manger for RabbitMQ

kadhikari opened the pull request:
When rabbitMQ is re-initialized and jormungandr is running,
jormungandr cannot push stat masage to rabbitMQ even if
the connection exists.

If it happens the connection will have to be re-initialized.

Commit:
jormungandr: Error handling in stat_manger for RabbitMQ

When rabbitMQ is re-initialized and jormungandr is running,
jormungandr cannot push stat masage to rabbitMQ even if
the connection exists.

If it happens the connection will have to be re-initialized.

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,17 @@ def fill_sections(self, stat_journey, resp_journey):
-        self.producer.publish(stat_request.SerializeToString())
+        try:
+            self.producer.publish(stat_request.SerializeToString())
+        except Exception as e:
```
too general

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,17 @@ def fill_sections(self, stat_journey, resp_journey):
+            if self.connection.connect():
+                self.connection.release()
+            #Initialize a new connection to RabbitMQ
+            self._init_rabbitmq()
```
we can use pool: http://ask.github.io/kombu/userguide/pools.html

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,17 @@ def fill_sections(self, stat_journey, resp_journey):
+                self.connection.release()
+            #Initialize a new connection to RabbitMQ
+            self._init_rabbitmq()
+            self.producer.publish(stat_request.SerializeToString())
```
if the connection fail before, this call raise a exception again

kadhikari replied:
Infact, the connection still exists mais stat_mpanager is not able to push message to rabbitMQ because of error "Broken pipe()??"

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,17 @@ def fill_sections(self, stat_journey, resp_journey):
+            logging.getLogger(__name__).info('Server went away, will be reconnected..')
 
+            #Release the existing connection and re-initialize rabbitMQ
+            if self.connection.connect():
```
we connect so we can close the connection? why?

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -97,6 +97,7 @@ def _init_rabbitmq(self):
             self.connection = kombu.Connection(self.broker_url)
             exchange = kombu.Exchange(self.exchange_name, type="direct")
             self.producer = self.connection.Producer(exchange=exchange)
+            self.save_stat = True
```
`_init_rabbitmq` seem to be call only if save_stat is true

kadhikari replied:
But if re-initialized, the variable save_stat has to be affected with "True" as false is affected by exception.

Commit:
jormungandr: Error handling modified after comments

If error of type connection or channel,
reinitialize the connection and if success
Push the stat message.

kadhikari commented:
Alex,
i modified the code after your comments. I couldn't find answer to some of your comments.
Could you verify my answers and corrections  please?

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,15 @@ def fill_sections(self, stat_journey, resp_journey):
+        try:
+            self.producer.publish(stat_request.SerializeToString())
+        except self.connection.connection_errors + self.connection.channel_errors:
+            logging.getLogger(__name__).info('Server went away, will be reconnected..')
```
```
 logging.getLogger(__name__).exception('Server went away, will be reconnected..')
```

this way we have the message of the exception and the stacktrace

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -336,8 +337,15 @@ def fill_sections(self, stat_journey, resp_journey):
+        except self.connection.connection_errors + self.connection.channel_errors:
+            logging.getLogger(__name__).info('Server went away, will be reconnected..')
+            #Initialize a new connection to RabbitMQ
+            self._init_rabbitmq()
```
we have to release/close the previous connection, if not we may leak handles

Commit:
jormungandr: Error handling in stat_manager corrected.

Message modified
Previous connection released.

kadhikari commented:
Alex,
i corrected for your last two comments.
